### ISSUE TITLE
Bulk insert breaks shared ArrayPool and System.Text.Json

### ISSUE THREAD

**Issue by akpl** [DarkWanderer/ClickHouse.Client#517]
It seems that batching inserts affects shared ArrayPool and consequently Json deserialization.
After `EnumerableExtensions.BatchRented<T>` is called, the arrays returned from Shared ArrayPool are multidimensional, which causes System.Text.Json to throw ArrayTypeMismatchException:
```
System.ArrayTypeMismatchException : Attempted to access an element as a type incompatible with the array.
   at System.Text.Json.Serialization.Converters.LargeObjectWithParameterizedConstructorConverter`1.InitializeConstructorArgumentCaches(ReadStack& state, JsonSerializerOptions options)
...
```
I'm not sure if this is not a bug in JsonSerializer but the issue appears after inserting bulk copy using ClickHouse.Client. It also breaks deserializing QueryStats inside the client so the library is affected.

Here is a test to reproduce:
```cs
[Test]
public async Task ShouldNotBreakArrayPool()
{
    var targetTable = "test." + SanitizeTableName($"array_pool");

    await connection.ExecuteStatementAsync($"DROP TABLE IF EXISTS {targetTable}");
    await connection.ExecuteStatementAsync($"CREATE TABLE IF NOT EXISTS {targetTable} (int Int32, str String, dt DateTime) ENGINE Null");

    using var bulkCopy = new ClickHouseBulkCopy(connection)
    {
        DestinationTableName = targetTable,
        BatchSize = 2
    };

    // This line works fine
    var pocOK = JsonSerializer.Deserialize<QueryStats>("{ }");

    await bulkCopy.InitAsync();
    await bulkCopy.WriteToServerAsync(Enumerable.Repeat(new object[] { 0, "a", DateTime.Now }, 1));

    // OPTION 1: now it throws ArrayTypeMismatchException
    var pocCRASH = JsonSerializer.Deserialize<QueryStats>("{ }");

    // OPTION 2: throws ArrayTypeMismatchException but hidden inside empty catch block (ClickHouse.Client.ADO.ClickHouseCommand.ExtractQueryStats)
    // var sql = $"SELECT COUNT(*) c FROM {targetTable}";
    // var rowNumber = await connection.ExecuteScalarAsync(sql);
}
```
I'm using `QueryStats` in above example but other classes deserialization can also break.

Version: 7.7.0
.NET Runtime: 6.0.26

**Comment by DarkWanderer** [comment]
Thank you for detailed analysis & report. This looks like an issue with JsonSerializer making assumptions about the contents of rented arrays - however equally, having contents it means there is a potential memory leak in BulkCopy. I will release fix for this in 7.7.1, let me know if it improves the situation

**Comment by akpl** [comment]
Thank you, I looked more into the issue and found the invalid conversion in `IntoBatches` method. Also updated the test to be more specific. Please take a look: https://github.com/DarkWanderer/ClickHouse.Client/pull/519

**Comment by DarkWanderer** [comment]
This is a very good catch - I was looking in a completely different direction

Merged your fix and released the fix as 7.7.1 - thank you very much for your contribution!
